src/DataTypes/Serializations/SerializationObject.cpp

```diff
@@ -334,12 +374,79 @@ void SerializationObject::deserializeBinaryBulkStatePrefix(
+        size_t num_tasks = std::min(settings.prefixes_deserialization_thread_pool->getMaxThreads(), structure_state_concrete->sorted_dynamic_paths.size());
+        std::vector<std::shared_ptr<DeserializationTask>> tasks;
+        tasks.reserve(num_tasks);
+        /// We need to create a copy of states cache for each task, because it's not thread-safe.
+        /// We will merge them together with the original cache later.
+        std::vector<std::unique_ptr<SubstreamsDeserializeStatesCache>> caches;
+        caches.reserve(num_tasks);
+
+        /// Create an entry for each dynamic path state beforehand.
+        for (const auto & path : structure_state_concrete->sorted_dynamic_paths)
+            object_state->dynamic_path_states[path] = nullptr;
+
+        /// All threads will use the same stream getter that is not thread safe.
+        std::mutex getter_mutex;
+        auto safe_getter = [&](const SubstreamPath & path)
+        {
+            std::unique_lock lock(getter_mutex);
+            return settings.getter(path);
+        };
+
+        size_t task_size = std::max(structure_state_concrete->sorted_dynamic_paths.size() / num_tasks, 1ul);
+        for (size_t i = 0; i != num_tasks; ++i)
+        {
+            auto cache_copy = cache ? std::make_unique<SubstreamsDeserializeStatesCache>(*cache) : nullptr;
+            size_t batch_start = i * task_size;
+            size_t batch_end = (i + 1) == num_tasks ? structure_state_concrete->sorted_dynamic_paths.size() : (i + 1) * task_size;
+            auto deserialize = [&, batch_start, batch_end, cache_ptr = cache_copy.get()]()
+            {
+                auto settings_copy = settings;
+                settings_copy.getter = safe_getter;
+                for (size_t j = batch_start; j != batch_end; ++j)
+                {
+                    settings_copy.path.push_back(Substream::ObjectDynamicPath);
+                    settings_copy.path.back().object_path_name = structure_state_concrete->sorted_dynamic_paths[j];
+                    dynamic_serialization->deserializeBinaryBulkStatePrefix(settings_copy, object_state->dynamic_path_states.at(structure_state_concrete->sorted_dynamic_paths[j]), cache_ptr);
+                    settings_copy.path.pop_back();
+                }
+            };
+
+            auto task = std::make_shared<DeserializationTask>(deserialize);
+            settings.prefixes_deserialization_thread_pool->scheduleOrThrow([task_ptr = task](){ task_ptr->tryExecute(); });
+            tasks.push_back(task);
+            caches.push_back(std::move(cache_copy));
+        }
+
+        /// Now when all tasks were put into thread pool, we can try to steal and execute some tasks.
+        for (const auto & task : tasks)
+            task->tryExecute();
+
+        /// Now all tasks are either executing by thread pool or were already executed by this thread.
+        /// Wait for all tasks to be executed.
+        for (const auto & task : tasks)
+            task->wait();
```

Comment:
In case of exception in one task we may still continue executing some task in `prefixes_deserialization_thread_pool`, but it may capture local variables. Should not we park all tasks explicitly in such case?

Reply:
Good suggestion, yes, let's wait for all tasks to be executed before rethrowing an exception.